man/podofopages.1
.TH "PODOFOPAGES" "1" "2010-12-09" "PoDoFo" "podofopages"
.PP
.SH NAME
podofopages \- move and delete pages in a PDF document
.PP
.SH SYNOPSIS
\fBpodofopages\fR [\-\-delete] [\-\-move] [inputfile] [outputfile]
.PP
.SH DESCRIPTION
.B podofopages
is one of the command line tools from the PoDoFo library that provide several
useful operations to work with PDF files\. It can move and delete pages in a
PDF document\.
.PP
.SH "OPTIONS"
.PP
\fB\-\-delete NUMBER\fR
.RS
.PP
Deletes the page NUMBER (number is 0\-based)\. The page will not really be
deleted from the PDF\. It is only removed from the so called pagestree and
therefore becomes hidden\. The content of the page can still be retrieved from
the document though\.
.RE
.PP
\fB\-\-move FROM TO\fR
.RS
.PP
Moves a page FROM TO in the document (FROM and TO are 0\-based)\.
.RE
.PP
.SH SEE ALSO
.BR podofobox (1),
.BR podofocolor (1),
.BR podofocountpages (1),
.BR podofocrop (1),
.BR podofoencrypt (1),
.BR podofogc (1),
.BR podofoimg2pdf (1),
.BR podofoimgextract (1),
.BR podofoimpose (1),
.BR podofoincrementalupdates (1),
.BR podofomerge (1),
.BR podofopdfinfo (1),
.BR podofotxt2pdf (1),
.BR podofotxtextract (1),
.BR podofouncompress (1),
.BR podofoxmp (1)
.PP
.SH AUTHORS
.PP
PoDoFo is written by Dominik Seichter <domseichter@web\.de> and others\.
.PP
This manual page was written by Oleksandr Moskalenko <malex@debian\.org> for
the Debian Project (but may be used by others)\.


ChangeLog
Version 0.8
	See SVN ChangeLog

Version 0.7
    Fixed PdfPainter::ArcTo
    Fixed crash in FlateDecode and LZWDecode predictor functions
    Fixed writing of unicode PdfStrings with brackets (will be escaped
	now)
    Fixed encoding of bytes in PdfName
    Added many new unit tests
    Added methods to free object memory of objects that are not needed
	anymore (These objects will be re-read from disk if they are
	needed again)
    Fixed a crash when appending PDFs
    Added unicode support on Win32 (wchar_t* constructors and methods
	where apropriate, e.g. file handling)
    Fixed a memory leak in PdfStream
    Small optimizations in various places
    Fixed DCTDecode filter for CMYK images
    PdfReference is now immutable, which allows for various optimizations
    Fixed PdfInputDevice::Read to return now the correct number of bytes
	read
    Fixed a memory leak in PdfImmediateWriter/PdfStreamedDocument
    Fixed several minor parsing issues
    Fixed adding text to existing pages
    Added Lua5.1 plan file support to podofoimpose
    Several fixes in creating XObjects from pages
    Fixed a possible crash in PdfNamesTree
    Added support for setting colors in PdfAnnotations
    Added advanced text drawing support to PdfPainter
    Fixed parsing of Type1 fonts
    Fixed deletion of PdfAnnations (fixes a memory leak)
    

Version 0.6
    Many many changes :) See SVN log

Version 0.5
    Added support for Embedded Files (annotations & named objects)
    Added support for ExtGStates when drawing
		initially only supports basic transparency
    Fixed reading values from nametrees
    Added support for named destinations
    Fixed a memory leak in PdfDestination::GetPage
    Pages do not know their page number inside of the document
    Fixed reading PdfActions from PDF files
    Moved filter implementations to PdfFiltersPrivate.h
    Added PdfFilter::CanEncode and PdfFilter::CanDecode
    Simpliefid PoDoFos handling of XRef tables
	
Version 0.4
	PdfImage now supports creating an image stream from a "raw bitmap"
		which can also be optionally Flate compressed
	Added some new Page-related methods to PdfDocument 
		Append - append one document to another
		InsertPages - insert a range of pages from one document to another
		DeletePages - delete a range of pages
	Added new tool podofomerge for merging two PDFs together
	Added methods to get & set a document's PageMode
	Added methods to set a document's FullScreen mode
	Added methods to set all the various ViewerPreferences for a document
	Added methods to set the document's PageLayout
	Added Outline support
		modified podofoinfo to display them, if present	
	Added a PdfDestination class
	Added PdfNamesTree class for handling the global named objects
	        modified podofoinfo to display them, if present
	PdfPainter can draw bezier curves
	Added XCode project for building on Mac OS X
	        fixed up conditionals in font code to enable building on
		MacOSX - but no font loading, YET
	Added support for writing linearized PDF
	Added support for garbage collection of unused objects
	Simplified PdfObject constructor
	Improved annotation support
	Added support to encode names and various name testcases
	Fixed ascent and descent of fonts
	Improved PdfImage API
	Added support for the creation of file identifiers which makes
		PoDoFo created PDF files work in more different PDF
		workflows
	PdfImage optionally takes ownership of buffers
	Fixed a major parser bug: The parser expected all objects in
		object streams to be of type dictionary.
	
Version 0.3
    TOTAL revamp of PdfObject & PdfVariant to enable clean/consistent object handling;
    	Added new PdfDocument object - new high level object for reading & writing a PDF doc;
	TOTAL revamp of the PdfDocument, PdfWriter & PdfParser relationship
		PdfDocument is now hub for both reading and writing a document
		it holds the PdfVecObjects - the others just reference them.
	TOTAL revamp of PdfPainter
		now uses PDF coordinates - UserUnits from bottom/left
			added PdfPainterMM for mm-based coords
		supports user-specified float precision
			and writes out floats in an optimal manner
		supports "appending" mode for drawing on existing documents
	Improved handling of the /Info dict for both reading and writing PDFs;
	Added new test app - podofopdfinfo, which will be used to dump metadata, etc. from a PDF;
	Added PdfError::DebugMessage() as the official way to write out debugging info;
		updated all other debugging msgs to use this;
	Added PdfError::DebugEnabled() to enable/disable display of debug messages;
	Added tracking of file size in PdfParser;
	Minor tweak to Linearization handling - to enable getting the status from a doc;
	Added getting GetPdfVersionAsString() to PdfWriter;
	Added new info/object getting methods to PdfDocument;
		bool IsLinearized();
		size_t FileSize();
		PdfObject* GetStructTreeRoot();
		PdfObject* GetMetadata();
		PdfObject* GetOutlines();
		PdfObject* GetAcroForm();
	Updated pdfinfo & podofopdfinfo to call the new PdfDocument methods;
	Added PdfDictionary and PdfArray classes;
	Added new PdfPagesTree (inside of PdfDocument.cpp) for handling walking a /Pages tree;
	Added new GetPageCount() method to PdfDocument;
	Modifications to PdfPage to attach it to a PdfDocument & construct from a PdfObject;
	Added new Legal and A3 standard page sizes;
	Changed page coordinates to be PDF unit-based instead of 1/1000mm;
	Changed PdfRect to use PDF units and also use bottom instead of top;
	Added ability to go between PdfRect and PdfArray & also get string version of a PdfRect;
	Added support for PdfPage to return all the standard boxes (Media, Crop, etc);
	Added support for fetching inherited values from pages (eg. boxes, rotation, etc.)
	Added more methods to PdfPage;
		GetRotation();
		GetNumAnnots();
	Use Exceptions now instead of error codes;
	Removed Init from PdfOutputDevice;
	Removed Init from PdfParser;
	Added LZW Filter support;

tools/podofopages/podofopages.cpp
/***************************************************************************
 *   Copyright (C) 2009 by Dominik Seichter                                *
 *   domseichter@web.de                                                    *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

#include <podofo.h>
#include <stdlib.h>

#include <iostream>
#include <iostream>
#include <string>
#include <stdexcept>
#include <vector>
#include <cstdio>

#include "DeleteOperation.h"
#include "MoveOperation.h"

using namespace PoDoFo;

#ifdef _HAVE_CONFIG
#include <config.h>
#endif // _HAVE_CONFIG

class BadConversion : public std::runtime_error {
public:
    BadConversion(const std::string& s)
        : std::runtime_error(s)
    { }
};

void print_help()
{
  printf("Usage: podofopages [inputfile] [outputfile]\n");
  printf("Options:\n");
  printf("\t--delete NUMBER\n");
  printf("\tDeletes the page NUMBER (number is 0-based)\n");
  printf("\tThe page will not really be deleted from the PDF.\n");
  printf("\tIt is only removed from the so called pagestree and\n");
  printf("\ttherefore invisible. The content of the page can still\n");
  printf("\tbe retrieved from the document though.\n\n");
  printf("\t--move FROM TO\n");
  printf("\tMoves a page FROM TO in the document (FROM and TO are 0-based)\n\n");
  printf("\nPoDoFo Version: %s\n\n", PODOFO_VERSION_STRING);
}

void work(const char* pszInput, const char* pszOutput, std::vector<Operation*> & rvecOperations)
{
    std::cout << "Input file: " << pszInput << std::endl;
    std::cout << "Output file: " << pszOutput << std::endl;

    PdfMemDocument doc(pszInput);

    int total = rvecOperations.size();
    int i = 1;
    std::vector<Operation*>::iterator it = rvecOperations.begin();
    while( it != rvecOperations.end() )
    {
        std::string msg = (*it)->ToString();
        std::cout << "Operation " << i << " of " << total << ": " << msg;

        (*it)->Perform( doc );

        ++it;
        ++i;
    }

    std::cout << "Operations done. Writing PDF to disk." << std::endl;

    doc.Write( pszOutput );

    std::cout << "Done." << std::endl;
}

double convertToInt(const std::string& s)
{
    std::istringstream i(s);
    int x;
    if (!(i >> x))
        throw BadConversion("convertToInt(\"" + s + "\")");
    return x;
}

int main( int argc, char* argv[] )
{
  char* pszInput = NULL;
  char* pszOutput = NULL;

  if( argc < 3 )
  {
    print_help();
    exit( -1 );
  }

  // Fill operations vector
  std::vector<Operation*> vecOperations;
  for( int i=1; i < argc; i++ )
  {
      std::string argument = argv[i];
      if( argument == "--delete" || argument == "-delete" )
      {
          int page = static_cast<int>(convertToInt( std::string(argv[i+1]) ));
          vecOperations.push_back( new DeleteOperation( page ) );
          ++i;
      }
      else if( argument == "--move" || argument == "-move" )
      {
          int from = static_cast<int>(convertToInt( std::string(argv[i+1]) ));
          int to = static_cast<int>(convertToInt( std::string(argv[i+2]) ));
          vecOperations.push_back( new MoveOperation( from, to ) );
          ++i;
          ++i;
      }
      else
      {
          if( pszInput == NULL )
          {
              pszInput = argv[i];
          }
          else if( pszOutput == NULL )
          {
              pszOutput = argv[i];
          }
          else
          {
              std::cerr << "Ignoring unknown argument: " << argument << std::endl;
          }
      }
  }

  if( !pszInput )
  {
      std::cerr << "Please specify an input file." << std::endl;
      exit( -2 );
  }

  if( !pszOutput )
  {
      std::cerr << "Please specify an output file." << std::endl;
      exit( -3 );
  }

  if( std::string(pszInput) == std::string(pszOutput) )
  {
      std::cerr << "Input and outpuf file must point to different files." << std::endl;
      exit( -4 );
  }

  try {
      work( pszInput, pszOutput, vecOperations );
  } catch( PdfError & e ) {
      std::cerr << "Error: An error " << e.GetError() << " ocurred." << std::endl;
      e.PrintErrorMsg();
      return e.GetError();
  } catch( std::runtime_error & re ) {
      std::cerr << "Error: An error " << re.what() << " ocurred." << std::endl;
      return -1;
  }

  // Delete operations vectore
  std::vector<Operation*>::iterator it = vecOperations.begin();
  while( it != vecOperations.end() )
  {
      delete (*it);
      ++it;
  }

  return 0;
}

